File: src/main/java/blockgame/block/TilePalette.java
package blockgame.block;

import java.util.ArrayList;
import java.util.List;

public class TilePalette {
    private static final List<TileState> _PALETTE = new ArrayList<>();
    public static TilePalette INSTANCE = new TilePalette();

    private TilePalette() {
        _PALETTE.add(new TileState(BlockRegistry.AIR));
    }

    public static int getTileRepresentation(TileState t) {
        if(_PALETTE.contains(t)) {
            return _PALETTE.indexOf(t);
        }
        _PALETTE.add(t);
        return _PALETTE.size()-1;
    }

    public static TileState getTileState(int t) throws Exception {
        if(_PALETTE.size() <= t) {
            throw new Exception("Invalid tile '$t'");
        }
        return _PALETTE.get(t);
    }
}

File: src/main/kotlin/block/TilePalette.kt
package block

object TilePalette {
    private var _palette: MutableList<TileState> = mutableListOf()

    init {
        _palette.add(TileState(BlockRegistration.AIR))
    }
    fun getTileRepresentation(t: TileState): Int {
        if(_palette.contains(t)) {
            return _palette.indexOf(t)
        }
        _palette.add(t)
        return _palette.size-1
    }

    fun getTileState(t: Int): TileState {
        if(_palette.size <= t) {
            throw Exception("Invalid tile '$t'")
        }
        return _palette[t]
    }
}